File: haskell/list-ops/src/ListOps.hs
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

import Data.Composition as Composition

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ xs [] = xs
foldl' f xs' (x:xs) = seq xs'' $ foldl' f xs'' xs
  where xs'' = f xs' x

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ xs [] = xs
foldr f xs (y:ys) = f y $ foldr f xs ys


length :: [a] -> Int
length = foldr (succ .: flip const) 0

reverse :: [a] -> [a]
reverse = foldl' reducer [] where
  reducer ys x = x : ys

map :: (a -> b) -> [a] -> [b]
map f = foldr reducer [] where
  reducer x ys = f x : ys

filter :: (a -> Bool) -> [a] -> [a]
filter test = foldr reducer [] where
  reducer x ys = if test x then x : ys else ys

(++) :: [a] -> [a] -> [a]
xs ++ [] = xs
[] ++ ys = ys
[x] ++ ys =  x:ys
xs ++ [y] = reverse $ y : reverse xs
(x:xs) ++ ys = seq (xs ++ ys) $ [x] ++ (xs ++ ys)


concat :: [[a]] -> [a]
concat = foldl' (++) []

File: typescript/list-ops/list-ops.ts
type Nullable<T> = T | null;
interface Reducer<A, E> {
    (acc: A, el: E): A
}

export default class List<T> {

    constructor(public values: Array<T> = []) {
    }
    isEmpty(): boolean {
        return this.values.length == 0
    }
    first(): Nullable<T> {
        return this.isEmpty() ? null : this.values[0]
    }
    rest(): List<T> {
        if (this.isEmpty())
            return this;
        const [_fst, ...rst] = this.values;
        return new List(rst);
    }
    append(l: List<T>): List<T> {
        if (l.isEmpty())
            return this;
        else
            return new List([...this.values, l.first() as T]).append(l.rest());
    }
    concat(l: List<List<T>>): List<T> {
        if (l.isEmpty())
            return this;
        else
            return this.append(l.first() as List<T>).concat(l.rest());
    }
    foldl<A>(reducer: Reducer<A, T>, acc: A): A {
        if (this.isEmpty())
            return acc;
        else
            return this.rest().foldl(reducer, reducer(acc, this.first() as T))
    }
    filter(predicate: (el: T) => boolean) {
        const filterReducer =
            (acc: List<T>, el: T) => predicate(el) ? acc.append(new List([el])) : acc
        return this.foldl(filterReducer, new List())
    }
    length() {
        return this.foldl((acc, _e) => acc + 1, 0)
    }
    map<R>(mapper: (el: T) => R) {
        const mapReducer = (acc: List<R>, el: T) => acc.append(new List([mapper(el)]))
        return this.foldl(mapReducer, new List())
    }
    reverse() {
        const reverseReducer = (acc: List<T>, el: T) => new List([el]).append(acc)
        return this.foldl(reverseReducer, new List())
    }
    last() {
        return this.reverse().first()
    }
    foldr<A>(reducer: Reducer<A, T>, acc: A): A {
        if (this.isEmpty())
            return acc;
        else {
            const beforeLast = this.reverse().rest().reverse()
            return beforeLast.foldr(reducer, reducer(acc, this.last() as T))
        }
    }
}